Data/ByteString/Builder/RealFloat.hs

```diff
@@ -0,0 +1,273 @@
+    c_ryu_d2s :: CDouble -> Ptr Word8 -> IO CInt
+
+foreign import ccall unsafe "static f2s_floating_decimal"
+    c_ryu_f2s_fd :: CFloat -> Ptr Word32 -> Ptr Int32 -> IO ()
+
+foreign import ccall unsafe "static d2s_floating_decimal"
+    c_ryu_d2s_fd :: CDouble -> Ptr Word64 -> Ptr Int32 -> IO ()
+
+foreign import ccall unsafe "static f2s_to_chars"
+    c_ryu_f2s_to_chars :: CUInt -> CInt -> CUChar -> Ptr Word8 -> IO CInt
+
+foreign import ccall unsafe "static d2s_to_chars"
+    c_ryu_d2s_to_chars :: CULong -> CInt -> CUChar -> Ptr Word8 -> IO CInt
+
+{-# INLINE ryu_f2s #-}
+ryu_f2s :: Float -> ByteString
+ryu_f2s f = unsafeDupablePerformIO $ do
+    fp <- mallocByteString 16 :: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_f2s (realToFrac f) p
+
+{-# INLINE ryu_d2s #-}
+ryu_d2s :: Double -> ByteString
+ryu_d2s f = unsafeDupablePerformIO $ do
+    fp <- mallocByteString 25 :: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_d2s (realToFrac f) p
+
+data FloatingDecimal64 = FD64 !Word64 !Int32
+data FloatingDecimal32 = FD32 !Word32 !Int32
+
+instance Show FloatingDecimal64 where
+  showsPrec p (FD64 m e) = showsPrec p m `mappend` showsPrec p '.' `mappend` showsPrec p e
+
+instance Show FloatingDecimal32 where
+  showsPrec p (FD32 m e) = showsPrec p m `mappend` showsPrec p '.' `mappend` showsPrec p e
+
+-- extracts base-10 converted mantissa and exponent using ryu algorithm
+-- NB: only valid if not NaN, +/-0, or +/-Inf
+{-# INLINE ryu_f2s_fd #-}
+ryu_f2s_fd :: Float -> FloatingDecimal32
+ryu_f2s_fd f = unsafeDupablePerformIO $
+    alloca $ \mOut -> do
+      alloca $ \eOut -> do
+        c_ryu_f2s_fd (realToFrac f) mOut eOut
+        m <- peek mOut
+        e <- peek eOut
+        return $ FD32 m e
+
+{-# INLINE ryu_d2s_fd #-}
+ryu_d2s_fd :: Double -> FloatingDecimal64
+ryu_d2s_fd f = unsafeDupablePerformIO $
+    alloca $ \mOut -> do
+      alloca $ \eOut -> do
+        c_ryu_d2s_fd (realToFrac f) mOut eOut
+        m <- peek mOut
+        e <- peek eOut
+        return $ FD64 m e
+
+asCBool :: Bool -> CUChar
+asCBool = fromIntegral . fromEnum
+
+{-# INLINE ryu_f2s_to_chars #-}
+ryu_f2s_to_chars :: Word32 -> Int32 -> Bool -> ByteString
+ryu_f2s_to_chars m e s = unsafeDupablePerformIO $ do
+    fp <- mallocByteString 16 :: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_f2s_to_chars (fromIntegral m) (fromIntegral e) (asCBool s) p
+
+{-# INLINE ryu_d2s_to_chars #-}
+ryu_d2s_to_chars :: Word64 -> Int32 -> Bool -> ByteString
+ryu_d2s_to_chars m e s = unsafeDupablePerformIO $ do
+    fp <- mallocByteString 25:: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_d2s_to_chars (fromIntegral m) (fromIntegral e) (asCBool s) p
+
+
+-- auxiliary fixed format printing functions
+decimalLength9:: Integral a => a -> Int
+decimalLength9 v
+  | v >= 100000000 = 9
```

Comment:
While I'm less concerned about truncation here, I would still recommend changing the signature from `Integral` to `FD32` for safety and perhaps efficiency.

Reply:
https://github.com/haskell/bytestring/pull/222/commits/96f30c7f12b781cbaa35a21345bb49b15b5b1191